pkg/controllermanager/controller/backupbucket/backup_bucket_actuator.go

```diff
@@ -0,0 +1,257 @@
+	"time"
+
+	"github.com/gardener/gardener/pkg/utils/kubernetes/health"
+
+	gardencorev1alpha1 "github.com/gardener/gardener/pkg/apis/core/v1alpha1"
+	gardencorev1alpha1helper "github.com/gardener/gardener/pkg/apis/core/v1alpha1/helper"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"github.com/gardener/gardener/pkg/operation/common"
+	"github.com/gardener/gardener/pkg/utils/flow"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+	"github.com/gardener/gardener/pkg/utils/retry"
+	"github.com/sirupsen/logrus"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kretry "k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	defaultTimeout  = 30 * time.Second
+	defaultInterval = 5 * time.Second
+)
+
+// Actuator acts upon BackupBucket resources.
+type Actuator interface {
+	// Reconcile reconciles the BackupBucket.
+	Reconcile(context.Context) error
+	// Delete deletes the BackupBucket.
+	Delete(context.Context) error
+}
+
+type actuator struct {
+	gardenClient client.Client
+	seedClient   client.Client
+	logger       logrus.FieldLogger
+	backupBucket *gardencorev1alpha1.BackupBucket
+}
+
+func newActuator(gardenClient, seedClient client.Client, bb *gardencorev1alpha1.BackupBucket, logger logrus.FieldLogger) Actuator {
+	return &actuator{
+		logger:       logger.WithField("backupbucket", bb.Name),
+		backupBucket: bb,
+		gardenClient: gardenClient,
+		seedClient:   seedClient,
+	}
+}
+
+func (a *actuator) Reconcile(ctx context.Context) error {
+	var (
+		g = flow.NewGraph("Backup Bucket Creation")
+
+		deployBackupBucketExtension = g.Add(flow.Task{
+			Name: "Deploying backup bucket extension resource",
+			Fn:   flow.TaskFn(a.deployBackupBucketExtension).RetryUntilTimeout(defaultInterval, defaultTimeout),
+		})
+
+		_ = g.Add(flow.Task{
+			Name:         "Waiting until backup bucket is reconciled",
+			Fn:           flow.TaskFn(a.waitUntilBackupBucketExtensionReconciled),
+			Dependencies: flow.NewTaskIDs(deployBackupBucketExtension),
+		})
+		f = g.Compile()
+	)
+	return f.Run(flow.Opts{
+		Logger:           a.logger,
+		ProgressReporter: a.reportBackupBucketProgress,
+		Context:          ctx,
+	})
+}
+
+func (a *actuator) Delete(ctx context.Context) error {
+	var (
+		g                  = flow.NewGraph("Backup bucket deletion")
+		deleteBackupBucket = g.Add(flow.Task{
+			Name: "Destroying backup bucket",
+			Fn:   flow.TaskFn(a.deleteBackupBucketExtension),
+		})
+		waitUntilBackupBucketExtensionDeleted = g.Add(flow.Task{
+			Name:         "Waiting until extension backup bucket is deleted",
+			Fn:           flow.TaskFn(a.waitUntilBackupBucketExtensionDeleted),
+			Dependencies: flow.NewTaskIDs(deleteBackupBucket),
+		})
+		_ = g.Add(flow.Task{
+			Name:         "Delete the secret referred by backup bucket extension",
+			Fn:           flow.TaskFn(a.deleteBackupBucketExtensionSecret),
```

Comment:
Why is this a separate task? It's also not a separate task in the reconciliation flow. Should we move it to the `deleteBackupBucketExtension` function? It's anyway not deleted unless the finalizer is released, right (I hope it has a finalizer)?

Reply:
No specific reason. Found it better while righting. Will combine it.

Finalizer... Didn't thought of it. gardener-extensions doesn't add finalizer on it. Will add Gardener Finalizer.